File: src/bmsx/machine/runtime/runtime_vblank.ts
import { IRQ_VBLANK } from '../bus/io';
import type { FrameState, Runtime } from './runtime';

export type RuntimeVblankSnapshot = {
	cyclesIntoFrame: number;
};

export class RuntimeVblankState {
	private clearBackQueuesAfterIrqWake = false;
	private haltIrqSignalSequence = 0;
	private haltIrqWaitArmed = false;
	private vblankSequence = 0;
	private lastCompletedVblankSequence = 0;
	private vblankCycles = 0;
	private vblankStartCycle = 0;
	private vblankActive = false;
	private frameStartCycle = 0;
	private activeTickCompleted = false;

	public get tickCompleted(): boolean {
		return this.activeTickCompleted;
	}

	public configureCycleBudget(runtime: Runtime): void {
		if (this.vblankCycles <= 0) {
			return;
		}
		if (this.vblankCycles > runtime.timing.cycleBudgetPerFrame) {
			throw new Error('Runtime fault: vblank_cycles must be less than or equal to cycles_per_frame.');
		}
		this.vblankStartCycle = runtime.timing.cycleBudgetPerFrame - this.vblankCycles;
		this.reset(runtime);
	}

	public setVblankCycles(runtime: Runtime, cycles: number): void {
		if (cycles <= 0) {
			throw new Error('Runtime fault: vblank_cycles must be greater than 0.');
		}
		if (cycles > runtime.timing.cycleBudgetPerFrame) {
			throw new Error('Runtime fault: vblank_cycles must be less than or equal to cycles_per_frame.');
		}
		this.vblankCycles = cycles;
		this.vblankStartCycle = runtime.timing.cycleBudgetPerFrame - this.vblankCycles;
		this.reset(runtime);
	}

	public getCyclesIntoFrame(runtime: Runtime): number {
		return runtime.machine.scheduler.nowCycles - this.frameStartCycle;
	}

	public resetScheduler(runtime: Runtime): void {
		runtime.machine.scheduler.reset();
		this.frameStartCycle = 0;
	}

	public reset(runtime: Runtime): void {
		this.resetScheduler(runtime);
		this.vblankActive = false;
		this.vblankSequence = 0;
		this.lastCompletedVblankSequence = 0;
		runtime.machine.inputController.sampleArmed = false;
		runtime.machine.irqController.postLoad();
		this.resetHaltIrqWait();
		runtime.machine.vdp.resetStatus();
		if (this.vblankStartCycle === 0) {
			this.setVblankStatus(runtime, true);
		}
		this.scheduleCurrentFrameTimers(runtime);
		runtime.refreshDeviceTimings(runtime.machine.scheduler.nowCycles);
	}

	public capture(runtime: Runtime): RuntimeVblankSnapshot {
		return {
			cyclesIntoFrame: this.getCyclesIntoFrame(runtime),
		};
	}

	public restore(runtime: Runtime, state: RuntimeVblankSnapshot): void {
		this.clearHaltUntilIrq(runtime);
		runtime.machineScheduler.reset();
		runtime.frameLoop.reset();
		runtime.screen.reset();
		this.resetScheduler(runtime);
		runtime.machine.scheduler.setNowCycles(state.cyclesIntoFrame);
		this.frameStartCycle = 0;
		this.vblankSequence = 0;
		this.lastCompletedVblankSequence = 0;
		this.activeTickCompleted = false;
		runtime.machine.irqController.postLoad();
		const vblankActive = (this.vblankStartCycle === 0)
			|| (this.getCyclesIntoFrame(runtime) >= this.vblankStartCycle);
		this.setVblankStatus(runtime, vblankActive);
		this.scheduleCurrentFrameTimers(runtime);
		runtime.refreshDeviceTimings(runtime.machine.scheduler.nowCycles);
	}

	public beginTick(): void {
		this.activeTickCompleted = false;
	}

	public abandonTick(): void {
		this.activeTickCompleted = false;
	}

	public handleBeginTimer(runtime: Runtime): void {
		if (!this.vblankActive) {
			this.enterVblank(runtime);
		}
	}

	public handleEndTimer(runtime: Runtime): void {
		if (this.vblankActive) {
			this.leaveVblank(runtime);
		}
		this.frameStartCycle = runtime.machine.scheduler.nowCycles;
		this.scheduleCurrentFrameTimers(runtime);
		if (this.vblankStartCycle === 0) {
			this.enterVblank(runtime);
		}
	}

	public clearHaltUntilIrq(runtime: Runtime): void {
		runtime.machine.cpu.clearHaltUntilIrq();
		this.resetHaltIrqWait();
		this.clearBackQueuesAfterIrqWake = false;
	}

	public consumeBackQueueClearAfterIrqWake(): boolean {
		if (!this.clearBackQueuesAfterIrqWake) {
			return false;
		}
		this.clearBackQueuesAfterIrqWake = false;
		return true;
	}

	public runHaltedUntilIrq(runtime: Runtime, state: FrameState): boolean {
		runtime.runDueTimers();
		if (!runtime.machine.cpu.isHaltedUntilIrq()) {
			this.resetHaltIrqWait();
			return false;
		}
		if (this.tryCompleteTickOnPendingVblankIrq(runtime, state)) {
			return true;
		}
		if (!this.haltIrqWaitArmed) {
			const pendingFlags = runtime.machine.irqController.pendingFlags();
			if (pendingFlags !== 0) {
				runtime.machine.cpu.clearHaltUntilIrq();
				return this.activeTickCompleted;
			}
			this.haltIrqSignalSequence = runtime.machine.irqController.signalSequence;
			this.haltIrqWaitArmed = true;
		}
		while (true) {
			if (runtime.machine.irqController.signalSequence !== this.haltIrqSignalSequence) {
				runtime.machine.cpu.clearHaltUntilIrq();
				this.resetHaltIrqWait();
				return this.activeTickCompleted;
			}
			if (state.cycleBudgetRemaining > 0) {
				const cyclesToTarget = runtime.machine.scheduler.nextDeadline() - runtime.machine.scheduler.nowCycles;
				if (cyclesToTarget <= 0) {
					runtime.runDueTimers();
					continue;
				}
				const idleCycles = cyclesToTarget < state.cycleBudgetRemaining ? cyclesToTarget : state.cycleBudgetRemaining;
				state.cycleBudgetRemaining -= idleCycles;
				runtime.advanceTime(idleCycles);
				if (this.tryCompleteTickOnPendingVblankIrq(runtime, state)) {
					return true;
				}
				continue;
			}
			return true;
		}
	}

	private scheduleCurrentFrameTimers(runtime: Runtime): void {
		runtime.machine.scheduler.scheduleVblankEnd(this.frameStartCycle + runtime.timing.cycleBudgetPerFrame);
		if (this.vblankStartCycle > 0 && this.getCyclesIntoFrame(runtime) < this.vblankStartCycle) {
			runtime.machine.scheduler.scheduleVblankBegin(this.frameStartCycle + this.vblankStartCycle);
		}
	}

	private setVblankStatus(runtime: Runtime, active: boolean): void {
		this.vblankActive = active;
		runtime.machine.vdp.setVblankStatus(active);
	}

	private enterVblank(runtime: Runtime): void {
		this.vblankSequence += 1;
		this.commitFrameOnVblankEdge(runtime);
		runtime.machine.inputController.onVblankEdge();
		this.setVblankStatus(runtime, true);
		runtime.machine.irqController.raise(IRQ_VBLANK);
		const frameState = runtime.frameLoop.currentFrameState;
		if (frameState !== null && this.isFrameBoundaryHalt(runtime)) {
			this.completeTickIfPending(runtime, frameState, this.vblankSequence);
			this.clearBackQueuesAfterIrqWake = true;
		}
	}

	private leaveVblank(runtime: Runtime): void {
		this.setVblankStatus(runtime, false);
	}

	private resetHaltIrqWait(): void {
		this.haltIrqWaitArmed = false;
		this.haltIrqSignalSequence = 0;
	}

	private tryCompleteTickOnPendingVblankIrq(runtime: Runtime, state: FrameState): boolean {
		if (!this.isFrameBoundaryHalt(runtime)) {
			return false;
		}
		if (this.vblankSequence === 0) {
			return false;
		}
		const pendingFlags = runtime.machine.irqController.pendingFlags();
		if ((pendingFlags & IRQ_VBLANK) === 0) {
			return false;
		}
		if (this.lastCompletedVblankSequence === this.vblankSequence) {
			return false;
		}
		this.completeTickIfPending(runtime, state, this.vblankSequence);
		this.clearBackQueuesAfterIrqWake = true;
		runtime.machine.cpu.clearHaltUntilIrq();
		this.resetHaltIrqWait();
		return true;
	}

	private isFrameBoundaryHalt(runtime: Runtime): boolean {
		return runtime.machine.cpu.getFrameDepth() === 1
			&& runtime.pendingCall === 'entry'
			&& runtime.machine.cpu.isHaltedUntilIrq();
	}

	private commitFrameOnVblankEdge(runtime: Runtime): void {
		runtime.machine.vdp.syncRegisters();
		runtime.machine.vdp.presentReadyFrameOnVblankEdge();
		runtime.machine.vdp.commitViewSnapshot();
	}

	private completeTickIfPending(runtime: Runtime, frameState: FrameState, vblankSequence: number): void {
		if (this.lastCompletedVblankSequence === vblankSequence) {
			return;
		}
		this.activeTickCompleted = true;
		runtime.machineScheduler.enqueueTickCompletion(runtime, frameState);
		this.lastCompletedVblankSequence = vblankSequence;
	}
}

File: src/bmsx_cpp/machine/runtime/runtime_vblank.cpp
#include "machine/runtime/runtime_vblank.h"

#include "core/engine_core.h"
#include "machine/bus/io.h"
#include "machine/runtime/runtime.h"
#include "machine/runtime/runtime_cpu_executor.h"
#include "machine/runtime/runtime_timing_config.h"
#include <algorithm>
#include <limits>
#include <stdexcept>
#include <string>

namespace bmsx {
namespace {
inline std::runtime_error runtimeFault(const std::string& message) {
	return BMSX_RUNTIME_ERROR("Runtime fault: " + message);
}
}

void RuntimeVblankState::configureCycleBudget(Runtime& runtime) {
	if (m_vblankCycles <= 0) {
		return;
	}
	if (m_vblankCycles > runtime.timing.cycleBudgetPerFrame) {
		throw runtimeFault("vblank_cycles must be less than or equal to cycles_per_frame.");
	}
	m_vblankStartCycle = runtime.timing.cycleBudgetPerFrame - m_vblankCycles;
	reset(runtime);
}

void RuntimeVblankState::setVblankCycles(Runtime& runtime, int cycles) {
	if (cycles <= 0) {
		throw runtimeFault("vblank_cycles must be greater than 0.");
	}
	if (cycles > runtime.timing.cycleBudgetPerFrame) {
		throw runtimeFault("vblank_cycles must be less than or equal to cycles_per_frame.");
	}
	m_vblankCycles = cycles;
	m_vblankStartCycle = runtime.timing.cycleBudgetPerFrame - m_vblankCycles;
	reset(runtime);
}

int RuntimeVblankState::getCyclesIntoFrame(const Runtime& runtime) const {
	return static_cast<int>(runtime.m_machine.scheduler().nowCycles() - m_frameStartCycle);
}

void RuntimeVblankState::resetScheduler(Runtime& runtime) {
	runtime.m_machine.scheduler().reset();
	m_frameStartCycle = 0;
}

void RuntimeVblankState::reset(Runtime& runtime) {
	resetScheduler(runtime);
	m_vblankActive = false;
	m_vblankSequence = 0;
	m_lastCompletedVblankSequence = 0;
	runtime.m_machine.inputController().restoreSampleArmed(false);
	runtime.m_machine.irqController().postLoad();
	resetHaltIrqWait();
	runtime.m_machine.vdp().resetStatus();
	if (m_vblankStartCycle == 0) {
		setVblankStatus(runtime, true);
	}
	scheduleCurrentFrameTimers(runtime);
	refreshDeviceTimings(runtime, runtime.m_machine.scheduler().nowCycles());
}

RuntimeVblankSnapshot RuntimeVblankState::capture(const Runtime& runtime) const {
	RuntimeVblankSnapshot state;
	state.cyclesIntoFrame = getCyclesIntoFrame(runtime);
	return state;
}

void RuntimeVblankState::restore(Runtime& runtime, const RuntimeVblankSnapshot& state) {
	clearHaltUntilIrq(runtime);
	runtime.machineScheduler.reset();
	runtime.frameLoop.reset();
	runtime.screen.reset();
	resetScheduler(runtime);
	runtime.m_machine.scheduler().setNowCycles(state.cyclesIntoFrame);
	m_frameStartCycle = 0;
	m_vblankSequence = 0;
	m_lastCompletedVblankSequence = 0;
	m_activeTickCompleted = false;
	runtime.m_machine.irqController().postLoad();
	const bool active = (m_vblankStartCycle == 0) || (getCyclesIntoFrame(runtime) >= m_vblankStartCycle);
	setVblankStatus(runtime, active);
	scheduleCurrentFrameTimers(runtime);
	refreshDeviceTimings(runtime, runtime.m_machine.scheduler().nowCycles());
}

void RuntimeVblankState::beginTick() {
	m_activeTickCompleted = false;
}

void RuntimeVblankState::abandonTick() {
	m_activeTickCompleted = false;
}

void RuntimeVblankState::handleBeginTimer(Runtime& runtime) {
	if (!m_vblankActive) {
		enterVblank(runtime);
	}
}

void RuntimeVblankState::handleEndTimer(Runtime& runtime) {
	if (m_vblankActive) {
		leaveVblank(runtime);
	}
	m_frameStartCycle = runtime.m_machine.scheduler().nowCycles();
	scheduleCurrentFrameTimers(runtime);
	if (m_vblankStartCycle == 0) {
		enterVblank(runtime);
	}
}

void RuntimeVblankState::clearHaltUntilIrq(Runtime& runtime) {
	runtime.m_machine.cpu().clearHaltUntilIrq();
	resetHaltIrqWait();
	m_clearBackQueuesAfterIrqWake = false;
}

bool RuntimeVblankState::consumeBackQueueClearAfterIrqWake() {
	if (!m_clearBackQueuesAfterIrqWake) {
		return false;
	}
	m_clearBackQueuesAfterIrqWake = false;
	return true;
}

bool RuntimeVblankState::runHaltedUntilIrq(Runtime& runtime, FrameState& frameState) {
	runDueRuntimeTimers(runtime);
	if (!runtime.m_machine.cpu().isHaltedUntilIrq()) {
		resetHaltIrqWait();
		return false;
	}
	if (tryCompleteTickOnPendingVblankIrq(runtime, frameState)) {
		return true;
	}
	if (!m_haltIrqWaitArmed) {
		const uint32_t pendingFlags = runtime.m_machine.irqController().pendingFlags();
		if (pendingFlags != 0u) {
			runtime.m_machine.cpu().clearHaltUntilIrq();
			return m_activeTickCompleted;
		}
		m_haltIrqSignalSequence = runtime.m_machine.irqController().signalSequence();
		m_haltIrqWaitArmed = true;
	}
	while (true) {
		if (runtime.m_machine.irqController().signalSequence() != m_haltIrqSignalSequence) {
			runtime.m_machine.cpu().clearHaltUntilIrq();
			resetHaltIrqWait();
			return m_activeTickCompleted;
		}
		if (frameState.cycleBudgetRemaining > 0) {
			const i64 cyclesToTarget = runtime.m_machine.scheduler().nextDeadline() - runtime.m_machine.scheduler().nowCycles();
			if (cyclesToTarget <= 0) {
				runDueRuntimeTimers(runtime);
				continue;
			}
			const int idleCycles = static_cast<int>(std::min<i64>(frameState.cycleBudgetRemaining, cyclesToTarget));
			frameState.cycleBudgetRemaining -= idleCycles;
			advanceRuntimeTime(runtime, idleCycles);
			if (tryCompleteTickOnPendingVblankIrq(runtime, frameState)) {
				return true;
			}
			continue;
		}
		return true;
	}
}

void RuntimeVblankState::scheduleCurrentFrameTimers(Runtime& runtime) {
	runtime.m_machine.scheduler().scheduleVblankEnd(m_frameStartCycle + runtime.timing.cycleBudgetPerFrame);
	if (m_vblankStartCycle > 0 && getCyclesIntoFrame(runtime) < m_vblankStartCycle) {
		runtime.m_machine.scheduler().scheduleVblankBegin(m_frameStartCycle + m_vblankStartCycle);
	}
}

void RuntimeVblankState::setVblankStatus(Runtime& runtime, bool active) {
	m_vblankActive = active;
	runtime.m_machine.vdp().setVblankStatus(active);
}

void RuntimeVblankState::enterVblank(Runtime& runtime) {
	m_vblankSequence += 1;
	commitFrameOnVblankEdge(runtime);
	runtime.m_machine.inputController().onVblankEdge();
	setVblankStatus(runtime, true);
	runtime.m_machine.irqController().raise(IRQ_VBLANK);
	if (runtime.frameLoop.frameActive && isFrameBoundaryHalt(runtime)) {
		completeTickIfPending(runtime, runtime.frameLoop.frameState, m_vblankSequence);
		m_clearBackQueuesAfterIrqWake = true;
	}
}

void RuntimeVblankState::leaveVblank(Runtime& runtime) {
	setVblankStatus(runtime, false);
}

void RuntimeVblankState::resetHaltIrqWait() {
	m_haltIrqWaitArmed = false;
	m_haltIrqSignalSequence = 0;
}

bool RuntimeVblankState::tryCompleteTickOnPendingVblankIrq(Runtime& runtime, FrameState& frameState) {
	if (!isFrameBoundaryHalt(runtime)) {
		return false;
	}
	if (m_vblankSequence == 0) {
		return false;
	}
	const uint32_t pendingFlags = runtime.m_machine.irqController().pendingFlags();
	if ((pendingFlags & IRQ_VBLANK) == 0u) {
		return false;
	}
	if (m_lastCompletedVblankSequence == m_vblankSequence) {
		return false;
	}
	completeTickIfPending(runtime, frameState, m_vblankSequence);
	m_clearBackQueuesAfterIrqWake = true;
	runtime.m_machine.cpu().clearHaltUntilIrq();
	resetHaltIrqWait();
	return true;
}

bool RuntimeVblankState::isFrameBoundaryHalt(Runtime& runtime) const {
	return runtime.m_machine.cpu().getFrameDepth() == 1
		&& runtime.m_pendingCall == Runtime::PendingCall::Entry
		&& runtime.m_machine.cpu().isHaltedUntilIrq();
}

void RuntimeVblankState::commitFrameOnVblankEdge(Runtime& runtime) {
	runtime.m_machine.vdp().syncRegisters();
	runtime.m_machine.vdp().presentReadyFrameOnVblankEdge();
	runtime.m_machine.vdp().commitViewSnapshot(*EngineCore::instance().view());
}

void RuntimeVblankState::completeTickIfPending(Runtime& runtime, FrameState& frameState, uint64_t vblankSequence) {
	if (m_lastCompletedVblankSequence == vblankSequence) {
		return;
	}
	m_activeTickCompleted = true;
	runtime.machineScheduler.enqueueTickCompletion(runtime, frameState);
	m_lastCompletedVblankSequence = vblankSequence;
}

} // namespace bmsx
